pkg/bucket/bandwidth/monitor.go

```diff
@@ -0,0 +1,186 @@
+
+	DoneCh          <-chan interface{} // Signal done to monitor
+	bucketAvgTicker *time.Ticker       // Ticker for calculating moving averages
+
+	pubsub *pubsub.PubSub
+}
+
+// NewMonitor returns a monitor with defaults.
+func NewMonitor() *Monitor {
+	m := &Monitor{
+		tokenizer:       token.NewNameTokenizer(),
+		activeBuckets:   make(map[int]*measurements.BucketMeasurement),
+		activeObjects:   make(map[int]*measurements.ObjectMeasurement),
+		Incoming:        make(chan Update),
+		DoneCh:          make(chan interface{}),
+		bucketAvgTicker: time.NewTicker(5 * time.Second),
+		pubsub:          pubsub.New(),
+	}
+	for i := 0; i < maxGoRoutines; i++ {
+		go m.listen()
+	}
+	return m
+}
+
+// GetReport gets the report for all bucket bandwidth details.
+func (m *Monitor) GetReport() *Report {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	return m.getReport()
+}
+
+func (m *Monitor) getReport() *Report {
+	report := &Report{
+		BucketStats: make(map[string]Details),
+	}
+	for bucket, bucketMeasurement := range m.activeBuckets {
+		bucketName, ok := m.tokenizer.GetName(bucket)
+		if ok {
+			report.BucketStats[bucketName] = Details{
+				AggregateBandwidth: bucketMeasurement.GetAvgMibps(),
+				MovingBandwidth:    bucketMeasurement.GetExpMovingAvgMibps(),
+			}
+		}
+		fmt.Printf("Bucket: %d, moving avg:%f aggregateAvg:%f\n", bucket, bucketMeasurement.GetExpMovingAvgMibps(), bucketMeasurement.GetAvgMibps())
+	}
+	return report
+}
+
+func (m *Monitor) listen() {
+	for {
+		select {
+		case <-m.bucketAvgTicker.C:
+			m.processAvg()
+		case <-m.DoneCh:
+			return
+		case sample := <-m.Incoming:
+			m.process(sample)
+		}
+	}
+}
+
+func (m *Monitor) getBucketMeasurement(bucketToken int, initTime time.Time) *measurements.BucketMeasurement {
+	bucketTracker, ok := m.activeBuckets[bucketToken]
+	if !ok {
+		bucketTracker = measurements.NewBucketMeasurement(initTime)
+		m.activeBuckets[bucketToken] = bucketTracker
+	}
+	return bucketTracker
+}
+
+func (m *Monitor) getObjectObjectMeasurement(objectToken int) *measurements.ObjectMeasurement {
+	objectTracker, ok := m.activeObjects[objectToken]
+	if !ok {
+		objectTracker = &measurements.ObjectMeasurement{}
+		m.activeObjects[objectToken] = objectTracker
+	}
+	return objectTracker
+}
+
+func (m *Monitor) process(update Update) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	bucketMeasurement := m.getBucketMeasurement(update.BucketToken, update.Start)
+	objectMeasurement := m.getObjectObjectMeasurement(update.ObjectToken)
+	objectMeasurement.UpdateMeasurement(update.Start, update.Stop, update.Bytes)
+	bucketMeasurement.IncrementBytes(update.Bytes)
+	if update.Done {
+		// TODO: Deleteme
+		fmt.Printf("Object Done: %d, moving avg:%f aggregateAvg:%f\n", update.ObjectToken, objectMeasurement.GetExpMovingAvgMibps(), objectMeasurement.GetAvgMibps())
+		delete(m.activeObjects, update.ObjectToken)
+		m.tokenizer.Release(update.ObjectToken)
+		// TODO: Implement archive of last n objects and buckets
```

Comment:
For objects that can be a lot of data.

Reply:
This is for per-server in-flight objects.